upnp_port_forward/tools/export.py

```diff
@@ -0,0 +1,84 @@
+import logging
+from typing import NamedTuple, Tuple
+
+import upnpclient
+
+from upnp_port_forward.exceptions import NoPortMapServiceFound
+
+
+class UPnPServiceNames(NamedTuple):
+    device_friendly_name: str
+    device_location: str
+    service_names: Tuple[str, ...]
+
+
+def fetch_add_portmapping_services() -> Tuple[UPnPServiceNames, ...]:
+    """
+    :return: returns the available devices and services for which the action 'AddPortMapping' exists
+    """
+    devices = upnpclient.discover()
+    if not devices:
+        raise NoPortMapServiceFound("No UPnP devices available")
+
+    services_with_AddPortMapping = []
+    for upnp_dev in devices:
+        service_names = []
+        for service in upnp_dev.services:
+            try:
+                service["AddPortMapping"]
+            except KeyError:
+                continue
+            else:
+                service_names.append(service.name)
+
+        if len(service_names) > 0:
+            services_with_AddPortMapping.append(
+                UPnPServiceNames(
+                    upnp_dev.friendly_name, upnp_dev.location, tuple(service_names)
+                )
+            )
+
+    if len(services_with_AddPortMapping) <= 0:
+        raise NoPortMapServiceFound(
+            "Unable to find a device with a port mapping service"
+        )
+
+    return tuple(services_with_AddPortMapping)
+
+
+def main() -> None:
+    logger = logging.getLogger("upnp_port_forward.tools.export")
+    try:
+        upnp_service_names_by_device = fetch_add_portmapping_services()
+        service_names_output = "\n"
+        for service in upnp_service_names_by_device:
+            service_names_output = service_names_output.join(
```

Comment:
This isn't going to produce the output that i think you are wanting.  In the event that there are multiple service names, the first pass will create one well formed block of text, joining the lines below with a newline.  However, on the second pass through the loop, since it continually uses `service_names_output.join(...)` it will interweave the well formed first one inbetween each of the line below.

I would suggest:

1. writing this as a function that you can pass the `upnp_service_names_by_device` list into, which will allow you to actually write a test for the output.
2. Building up the parts into a list and then joining them all at the end.

Reply:
>This isn't going to produce the output that i think you are wanting. In the event that there are multiple service names, the first pass will create one well formed block of text, joining the lines below with a newline. However, on the second pass through the loop, since it continually uses service_names_output.join(...) it will interweave the well formed first one inbetween each of the line below.

Thank you for pointing this out ! :sweat_smile: I changed accordingly.